### PR TITLE
Fix #769 Handle deconstruction assignation for undefined variable

### LINKED ISSUE #769: Array deconstructing reports new variables as "avoid using undefined variables"
- PHPMD version: 2.8.1
- PHP Version: 7.4.5
- Installation type: composer
- Operating System / Distribution & Version: Mac OSX 10.14.2

### Current Behavior
```php
[$foo, $bar] = split();
```

returns with errors like `Avoid using undefined variables such as $foo which will  lead to PHP notices`.

I tried this with a SCCE that only called a simple method, but curiously that one did not fail. Only after I put the method in a class i was able to reproduce it.

### Expected Behavior
Since it's a new variable from array deconstruction I would expect no error.

### Steps To Reproduce:
1. Create test.php:
   ```php
   <?php

   class Test
   {
       private static function split($string)
       {
           return [$string, $string];
       }

       public function deconstruct()
       {
           [$foo, $bar] = self::split('foo bar');
       }
   }
   ```
2. Run phpmd: `vendor/bin/phpmd test.php text phpmd.xml`
   Here's my config:
   ```xml
   <?xml version="1.0"?>
   <ruleset name="phpcstd" xmlns="https://pmd.sourceforge.io/ruleset_xml_schema.xsd">
       <rule ref="rulesets/codesize.xml"/>
       <rule ref="rulesets/design.xml"/>
       <rule ref="rulesets/naming.xml"/>
       <rule ref="rulesets/cleancode.xml"/>
   </ruleset>
```
3. See the following errors:
   ```
   /xxxx/test.php:12	Avoid using undefined variables such as '$foo' which will lead to PHP notices.
   /xxxx/test.php:12	Avoid using undefined variables such as '$bar' which will lead to PHP notices.
   ```

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

### PR DESCRIPTION
**Pull request by kylekatarnls**
Type: bugfix
Issue: Fix #769
Breaking change: no

### CODE REVIEW

**Comment by kylekatarnls** [comment]
I tested this with:
```php
public function testRuleDoesNotApplyToDeconstruct()
{
    $rule = new UndefinedVariable();
    $rule->setReport($this->getReportMock(0));
    $rule->apply($this->getMethod());
 }
```

But I didn't add it because #756 will handle those kind of test globally.

**Comment by kylekatarnls** [comment]
@spaceemotion Maybe you can test this patch on your case and give your feedback.

File: src/test/resources/files/Rule/CleanCode/UndefinedVariable/testRuleDoesNotApplyToDeconstruct.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * For full copyright and license information, please see the LICENSE file.
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @author Manuel Pichler <mapi@phpmd.org>
+ * @copyright Manuel Pichler. All rights reserved.
+ * @license https://opensource.org/licenses/bsd-license.php BSD License
+ * @link http://phpmd.org/
+ */
+
+class testRuleDoesNotApplyToDeconstruct
```
Comment (ravage84) [inline src/test/resources/files/Rule/CleanCode/UndefinedVariable/testRuleDoesNotApplyToDeconstruct.php:18]: Is there a reason why you did not name it ``testRuleDoesNotApplyToArrayDeconstruct``, mentioning the ``Array``?

Comment (ravage84) [inline src/test/resources/files/Rule/CleanCode/UndefinedVariable/testRuleDoesNotApplyToDeconstruct.php:18]: Is this test file used anywhere? :confused:

**Comment by ravage84** [comment]
> I tested this with:
> 
> ```
> public function testRuleDoesNotApplyToDeconstruct()
> {
>     $rule = new UndefinedVariable();
>     $rule->setReport($this->getReportMock(0));
>     $rule->apply($this->getMethod());
>  }
> ```
> 
> But I didn't add it because #756 will handle those kind of test globally.

@kylekatarnls I would prefer to keep PRs self containing, even if the merge of another one could change things.

**Comment by kylekatarnls** [comment]
It's self-contained: the test is present, it just depends on a other PR to be able to run it. If I do the contrary, we will have to fix a conflict then re-remove the extra code after #756. So the easy path is to review #756 first and get rid of the problem.

File: src/test/resources/files/Rule/CleanCode/UndefinedVariable/testRuleDoesNotApplyToDeconstruct.php (hunk 1 above)
Comment (kylekatarnls) [reply src/test/resources/files/Rule/CleanCode/UndefinedVariable/testRuleDoesNotApplyToDeconstruct.php:18]: None are correct, I will change for the PHP documentation wording: Symmetric array destructuring

**Comment by spaceemotion** [comment]
@kylekatarnls can confirm that this version fixes the error with the `test.php` file I've posted in my issue 🎉

**Pull request merged**